### PR TITLE
DISTPG-364 Backup and disaster recovery

### PR DESCRIPTION
**Pull request by nastena1606**
modified:   docs/solutions/backup-recovery.md
	modified:   docs/solutions/dr-pgbackrest-setup.md

### CODE REVIEW

**Review by patrickbirch — APPROVED** [review]
LGTM

File: docs/solutions/backup-recovery.md [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+## Overview
+
+A Disaster Recovery (DR) solution ensures that a system can be quickly restored to a normal operational state if something unexpected happens. When operating a database, you would back up the data as frequently as possible and have a mechanism to restore that data when needed. Disaster Recovery is often mistaken for high availability (HA), but they are two different concepts altogether:
+
+- High availability ensures guaranteed service levels at all times. This solution involves configuring one or more standby systems to an active database, and the ability to switch seamlessly to that standby when the primary database becomes unavailable, for example, during a power outage or a server crash. To learn more about high-availability solutions with Percona Distribution for PostgreSQL, refer to [High Availability in PostgreSQL with Patroni](high-availability.md).
+- Disaster Recovery protects the database instance against accidental or malicious data loss or data corruption. Disaster recovery can be achieved by using either the options, provided by PostgreSQL, or external extensions. 
+
+    ??? "PostgreSQL disaster recovery options"
```
Comment (hqakhtar) [inline docs/solutions/backup-recovery.md]: I presume that "???" is either a typo or some unprintable characters that you might want to remove.

Also, this should probably be a subheading.

File: docs/solutions/backup-recovery.md [hunk 2]
```diff
@@ -0,0 +1,81 @@
+# Backup and disaster recovery in Percona Distribution for PostgreSQL
+
+## Overview
+
+A Disaster Recovery (DR) solution ensures that a system can be quickly restored to a normal operational state if something unexpected happens. When operating a database, you would back up the data as frequently as possible and have a mechanism to restore that data when needed. Disaster Recovery is often mistaken for high availability (HA), but they are two different concepts altogether:
+
+- High availability ensures guaranteed service levels at all times. This solution involves configuring one or more standby systems to an active database, and the ability to switch seamlessly to that standby when the primary database becomes unavailable, for example, during a power outage or a server crash. To learn more about high-availability solutions with Percona Distribution for PostgreSQL, refer to [High Availability in PostgreSQL with Patroni](high-availability.md).
+- Disaster Recovery protects the database instance against accidental or malicious data loss or data corruption. Disaster recovery can be achieved by using either the options, provided by PostgreSQL, or external extensions.
```
Comment (hqakhtar) [inline docs/solutions/backup-recovery.md]: Remove the "," after options IMHO

File: docs/solutions/backup-recovery.md [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+- High availability ensures guaranteed service levels at all times. This solution involves configuring one or more standby systems to an active database, and the ability to switch seamlessly to that standby when the primary database becomes unavailable, for example, during a power outage or a server crash. To learn more about high-availability solutions with Percona Distribution for PostgreSQL, refer to [High Availability in PostgreSQL with Patroni](high-availability.md).
+- Disaster Recovery protects the database instance against accidental or malicious data loss or data corruption. Disaster recovery can be achieved by using either the options, provided by PostgreSQL, or external extensions. 
+
+    ??? "PostgreSQL disaster recovery options"
+
+        PostgreSQL offers multiple options for setting up database disaster recovery. 
+
+        - **[pg_dump](https://www.postgresql.org/docs/11/app-pgdump.html) or the [pg_dumpall](https://www.postgresql.org/docs/11/app-pg-dumpall.html) utilities**
```
Comment (hqakhtar) [inline docs/solutions/backup-recovery.md]: IMHO, we should be providing links to the latest PG documentation. Right now, some links point to 11 and some to current.

File: docs/solutions/backup-recovery.md [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+        This method involves backing up the PostgreSQL data directory to a different location, and restoring it when needed. 
+
+        | Advantages   | Disadvantages   |
+        | ------------ | --------------- |
+        | Consistent snapshot of the data directory or the whole data disk volume | 1. Requires stopping PostgreSQL in order to copy the files. This is not practical for most production setups.<br> 2. No backup of individual databases or tables.|
+
+        - **PostgreSQL [pg_basebackup](https://www.postgresql.org/docs/11/app-pgbasebackup.html)**
+
+        This backup tool is provided by PostgreSQL. It uses the [PostgreSQL Write Ahead logs](https://www.postgresql.org/docs/current/wal-intro.html) (WAL) and can thus back up data even when the database instance is running.
```
Comment (hqakhtar) [inline docs/solutions/backup-recovery.md]: This is factually incorrect. Please see the pg_basebackup documentation

File: docs/solutions/backup-recovery.md [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+`pgBackRest` can restore a database to a specific point in time in the past. This is the case where a database is not inaccessible but perhaps contains corrupted data. Using the point-in-time recovery, a database administrator can restore the database to the last known good state.  
+
+Finally, `pgBackRest` also supports restoring PostgreSQL databases to a different PostgreSQL instance or a separate data directory.
+
+
+## System architecture
+
+The following diagram illustrates the architecture layout:
+
+![pgBackRest Implementation Architecture](../_images/diagrams/DR-architecture.png)
```
Comment (hqakhtar) [inline docs/solutions/backup-recovery.md]: This link seems to be broken

File: docs/solutions/backup-recovery.md [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+When it comes to restoring, `pgBackRest` can do a full or a delta restore. A _full_ restore needs an empty PostgreSQL target directory. A _delta_ restore is intelligent enough to recognize already-existing files in the PostgreSQL data directory, and update only the ones the backup contains. 
+
+`pgBackRest` supports remote repository hosting and can even use cloud-based services like AWS S3, Google Cloud Services Cloud Storage, Azure Blob Storage for saving backup files. It supports parallel backup through multi-core processing and compression. By default, backup integrity is verified through checksums, and saved files can be encrypted for enhanced security.
+
+`pgBackRest` can restore a database to a specific point in time in the past. This is the case where a database is not inaccessible but perhaps contains corrupted data. Using the point-in-time recovery, a database administrator can restore the database to the last known good state.  
+
+Finally, `pgBackRest` also supports restoring PostgreSQL databases to a different PostgreSQL instance or a separate data directory.
+
+
+## System architecture
```
Comment (hqakhtar) [inline docs/solutions/backup-recovery.md]: This seems a little too abrupt. Perhaps some leading text explaining what this section is for will be helpful.

File: docs/solutions/backup-recovery.md [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+
+The architecture consists of three server instances:
+
+- `pg-primary` hosts the primary PostgreSQL server. Note that "primary" here means the main database instance and does not refer to the primary node of a PostgreSQL replication cluster or a HA setup.
+- `pg-repo` is the remote backup repository and hosts `pgBackRest`. It’s important to host the backup repository on a physically separate instance, to be accessed when the target goes down. 
+- `pg-secondary` is the _secondary_ PostgreSQL node. Don't confuse it with a hot standby. "Secondary" in this context means a PostgreSQL instance that’s idle. We will restore the database backup to this instance when the primary PostgreSQL instance goes down. 
+
+In our architecture, `pgBackRest` is located on a remote host and the nodes communicate with each other via passwordless SSH.
+
+!!! note
```
Comment (hqakhtar) [inline docs/solutions/backup-recovery.md:86]: Is "!!!" expected here?

File: docs/solutions/dr-pgbackrest-setup.md [hunk 8]
```diff
@@ -0,0 +1,481 @@
+# Deploying backup and disaster recovery solution in Percona Distribution for PostgreSQL
+
+This document provides instructions of how to set up and test the backup and disaster recovery solution in Percona Distribution for PostgreSQL with `pgBackRest`. For technical overview and architecture description of this solution, refer to [Backup and disaster recovery in Percona Distribution for PostgreSQL](backup-recovery.md).
+
+## Deployment 
+
+As the example configuration, we will use the nodes with the following IP addresses:
+
+| **Node name**   | **Internal IP address** |
```
Comment (hqakhtar) [inline docs/solutions/dr-pgbackrest-setup.md:9]: Would be nice to have the same ordering of these nodes as mentioned in the "Components" section.

File: docs/solutions/backup-recovery.md [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+
+### Components:
+
+The architecture consists of three server instances:
+
+- `pg-primary` hosts the primary PostgreSQL server. Note that "primary" here means the main database instance and does not refer to the primary node of a PostgreSQL replication cluster or a HA setup.
+- `pg-repo` is the remote backup repository and hosts `pgBackRest`. It’s important to host the backup repository on a physically separate instance, to be accessed when the target goes down. 
+- `pg-secondary` is the _secondary_ PostgreSQL node. Don't confuse it with a hot standby. "Secondary" in this context means a PostgreSQL instance that’s idle. We will restore the database backup to this instance when the primary PostgreSQL instance goes down. 
+
+In our architecture, `pgBackRest` is located on a remote host and the nodes communicate with each other via passwordless SSH.
```
Comment (hqakhtar) [inline docs/solutions/backup-recovery.md]: We should also add a security warning here that this passwordless SSH configuration may not be ideal for your environment. Please choose one that satisfies your organization's/client's security requirements.

File: docs/solutions/dr-pgbackrest-setup.md [hunk 10]
```diff
@@ -0,0 +1,481 @@
...
+   sudo apt-get install percona-pgbackrest
+   ```
+
+* On RHEL / CentOS:
+
+   ``` bash
+   sudo yum install percona-pgbackrest
+   ```
+
+### Create the `pgBackRest` configuration file
```
Comment (hqakhtar) [inline docs/solutions/dr-pgbackrest-setup.md:205]: This section is asking the user to perform far too many operations that should really be handled by our packages.

**Review by hqakhtar — COMMENTED** [review]
Other than the inline feedback, here is some additional feedback:

Files to be removed/fixed:
- ha-setup-yum.md file shouldn't be included in this PR
- mkdocs-base.yml has mentions of 11.14. These should be included in a separate PR.

Overall, this seems like a fairly basic tutorial on pgbackrest. It doesn't provide any insights into any pitfalls, configuration parameters, or best practices even.

File: docs/solutions/backup-recovery.md (hunk 1 above)
Comment (nastena1606) [reply docs/solutions/backup-recovery.md]: "???" is the formatting for the collapsible section, collapsed by default. 

Amended padding to make it a separate subsection

File: docs/solutions/backup-recovery.md (hunk 3 above)
Comment (nastena1606) [reply docs/solutions/backup-recovery.md]: Fixed

File: docs/solutions/backup-recovery.md (hunk 4 above)
Comment (nastena1606) [reply docs/solutions/backup-recovery.md]: Amended

File: docs/solutions/backup-recovery.md (hunk 5 above)
Comment (nastena1606) [reply docs/solutions/backup-recovery.md]: This is the markdown formatting. The image is a separate file and renders in the html output. You can see the test deployment here: https://new.percona.com/doc/postgresql/11/solutions/backup-recovery.html (SSO login is required)

File: docs/solutions/backup-recovery.md (hunk 7 above)
Comment (nastena1606) [reply docs/solutions/backup-recovery.md:86]: Yes, it marks a note and will be rendered as an admonition in the output.

File: docs/solutions/dr-pgbackrest-setup.md (hunk 8 above)
Comment (nastena1606) [reply docs/solutions/dr-pgbackrest-setup.md:9]: Fixed

**Comment by nastena1606** [comment]
@EngineeredVirus, thank you for the feedback. The source for this document is the article provided by outsource authors and confirmed internally before it was to be added to the docs.
I've amended the docs to address your comments, please review.

File: docs/solutions/backup-recovery.md (hunk 6 above)
Comment (nastena1606) [reply docs/solutions/backup-recovery.md]: Added an intro section

File: docs/solutions/backup-recovery.md (hunk 9 above)
Comment (nastena1606) [reply docs/solutions/backup-recovery.md]: Added

File: docs/solutions/backup-recovery.md (hunk 2 above)
Comment (nastena1606) [reply docs/solutions/backup-recovery.md]: Done

**Review by hqakhtar — APPROVED** [review]
LGTM

**Pull request merged**